src/mlpack/core/data/load_csv_impl.hpp

```diff
@@ -0,0 +1,222 @@
+ * In this mlpack's version, all the arma dependencies were removed or replaced
+ * accordingly, making the parser totally independent of armadillo.
+ *
+ * This parser will be totally independent to any linear algebra library.
+ * This can be used to load data into any matrix, i.e. arma and bandicoot
+ * in future.
+ *
+ * https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.5.x/include/armadillo_bits/diskio_meat.hpp
+ * Copyright 2008-2016 Conrad Sanderson (http://conradsanderson.id.au)
+ * Copyright 2008-2016 National ICT Australia (NICTA)
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_CSV_PARSER_IMPL_HPP
+#define MLPACK_CORE_DATA_CSV_PARSER_IMPL_HPP
+
+#include "load_csv.hpp"
+
+namespace mlpack
+{
+namespace data
+{
+  /**
+   * Given the address of a martix element(val)
+   * sets it equal to the provided value(token)
+   * example calling: convert_token<eT>(x.at(row, col), token)
+   */
+  template<typename MatType>
+  bool LoadCSV::ConvertToken(typename MatType::elem_type& val, const std::string& token)
+  {
+    const size_t N = size_t(token.length());
+
+    if (N == 0)
+    {
+      val = typename MatType::elem_type(0);
+      return true;
+    }
+
+    const char* str = token.c_str();
+
+    if ((N == 3) || (N == 4))
+    {
+      const bool neg = (str[0] == '-');
+      const bool pos = (str[0] == '+');
+
+      const size_t offset = ((neg || pos) && (N == 4)) ? 1 : 0;
+
+      const char sig_a = str[offset];
+      const char sig_b = str[offset+1];
+      const char sig_c = str[offset+2];
+
+      if (((sig_a == 'i') || (sig_a == 'I')) &&
+          ((sig_b == 'n') || (sig_b == 'N')) &&
+          ((sig_c == 'f') || (sig_c == 'F')))
+      {
+        val = neg ? -(std::numeric_limits<typename MatType::elem_type>::infinity()) :
+            std::numeric_limits<typename MatType::elem_type>::infinity();
+        return true;
+      }
+      else if (((sig_a == 'n') || (sig_a == 'N')) &&
+         ((sig_b == 'a') || (sig_b == 'A')) &&
+         ((sig_c == 'n') || (sig_c == 'N')))
```

Comment:
We can delay the conversion to `c_str` and convert the `token` to lowercase using
```
std::transform(token.begin(), token.end(), token,begin(), ::tolower);
```

This will eliminate the requirement of `sig_*` and we can simply do `token.substr(1, 3) == "inf"` and `token.substr(1, 3) == "nan"`.

After these, we can then convert the `token` to `c_str` for later use.

Reply:
@RishabhGarg108 I'm not sure that using `std::transform()` and `.substr()` is the most efficient way to do this.  Bear in mind that `ConvertToken()` needs to be very quick as it needs to process each token in a CSV file, which may be millions of times while processing one file.

Calling `std::transform` creates an overhead due to requiring a writeable copy of the token (where the copy will probably require memory allocation, which is slow), as well as looping through the characters. Using `.substr()` creates further loops.

This is in contrast to what's done now, where the relevant characters are simply extracted (via `sig_a` etc) and compared to a small fixed number of lower and upper characters.  No string copies, no explicit memory allocation, and no loops, so there is very little overhead.

Reply:
Ahh, I see. Thanks for the explanation! I overlooked the fact that this function is called a huge number of times. Anyways, the suggestion was just to simplify the code, so we can definitely live with it if it's fast :)